File: NC/longest-repeating-substring-with-replacement.cpp
class Solution
{
public:
    char getNewMaxNum(unordered_map<char, int> &charCount)
    {
        char ans = charCount.begin()->first;

        for (auto i : charCount)
        {
            if (i.second > charCount[ans])
            {
                ans = i.first;
            }
        }
        return ans;
    }

    int characterReplacement(string s, int k)
    {
        unordered_map<char, int> charCount;
        if (s.length() < 1)
        {
            return 0;
        }
        int ans = 1;
        charCount[s[0]] = 1;
        char maxNum = s[0];

        for (int left = 0, right = 1; right < s.length(); right++)
        {
            charCount[s[right]]++;
            while (right - left + 1 > charCount[maxNum] + k)
            {
                charCount[s[left]]--;

                left++;
                maxNum = getNewMaxNum(charCount);
            }
            cout << "left is " << left << ", right is " << right << endl;
            ans = max(ans, right - left + 1);
            maxNum = getNewMaxNum(charCount);
        }

        return ans;
    }
};

File: NC/java/LongestRepeatingSubstringWithReplacement.java
package NC.java;

import java.util.HashMap;
import java.util.Map;

public class LongestRepeatingSubstringWithReplacement {
    private char getMaxChar(Map<Character, Integer> charCount) {
        char ans = 'a';

        for (var pair : charCount.entrySet()) {
            if (pair.getValue() > charCount.getOrDefault(ans, 0)) {
                ans = pair.getKey();
            }
        }

        return ans;
    }

    public int characterReplacement(String s, int k) {

        if (s.length() == 0) {
            return 0;
        }

        Map<Character, Integer> charCount = new HashMap();

        charCount.put(s.charAt(0), 1);

        int ans = 1;

        for (int i = 0, j = 1; j < s.length(); j++) {
            char rightChar = s.charAt(j);
            charCount.put(rightChar, charCount.getOrDefault(rightChar, 0) + 1);

            while (j - i + 1 - charCount.get(this.getMaxChar(charCount)) > k) {
                char leftChar = s.charAt(i);
                charCount.put(leftChar, charCount.get(leftChar) - 1);
                i++;
            }

            ans = Math.max(ans, j - i + 1);

        }

        return ans;
    }
}
